src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -295,6 +315,27 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
         continue;
       }
 
+      if (sched_cls_info.running_tasks.size() >= sched_cls_info.capacity) {
+        RAY_LOG(DEBUG) << "Hit cap! time=" << get_time_()
+                       << " next update time=" << sched_cls_info.next_update_time;
+        if (get_time_() >= sched_cls_info.next_update_time) {
+          // Don't increase the capacity on the very first update since that
+          // will unblock a new task immediately, instead of after the timeout.
+          if (sched_cls_info.num_updates > 0) {
+            sched_cls_info.capacity++;
+          }
+
+          int64_t wait_time =
+              (1e6 * sched_cls_cap_interval_ms_) * (1L << sched_cls_info.num_updates++);
+          sched_cls_info.next_update_time = get_time_() + wait_time;
```

Comment:
I feel next_update_time always increases actually. Do we have the reset there? 

What if the cap is 2 and it eventually it increased to 8, does this mean, later for this scheduling class, we don't wait if there are less than 8 tasks?


Should we do things like (current_capacity - allowed_capacity) and when scheduling tasks exceeds the allowed capacity, we wait the seconds exponential to the difference?

Reply:
it gets reset when all the tasks in the scheduling class finish and the entry gets removed. There's a c++ unit test for this.

Reply:
I like the suggestion of making this based on (current - allowed), which eliminates the need for reset, can we implement that?